src/runtime_src/core/edge/skd/sk_daemon.cpp

```diff
@@ -203,13 +243,55 @@ static void softKernelLoop(char *name, char *path, uint32_t cu_idx)
       break;
     }
 
+    syslog(LOG_INFO, "Got new kernel command!\n");
+    
     /* Reg file indicates the kernel should not be running. */
     if (!(args_from_host[0] & 0x1))
       continue; //AP_START bit is not set; New Cmd is not available
 
-    /* Start run the soft kernel. */
-    kernel_return = kernel(&args_from_host[1], &ops);
+    // New PS Kernel implementation
+    // Check for call signature of only 2 arguments and 2nd argument has name of ops
+    if((args.size()==2 && args[1].name.compare("ops")==0) || args.empty()) {
+    } else {
+      // FFI PS Kernel implementation
+      // Map buffers used by kernel
+      for(int i=0;i<args.size();i++) {
+	if(args[i].type == xrt_core::pskernel::kernel_argument::argtype::global) {
+	  uint64_t *buf_addr_ptr = (uint64_t*)(&args_from_host[args[i].offset/4]);
+	  uint64_t buf_addr = reinterpret_cast<uint64_t>(*buf_addr_ptr);
+	  uint64_t *buf_size_ptr = (uint64_t*)(&args_from_host[args[i].offset/4+2]);
+	  uint64_t buf_size = reinterpret_cast<uint64_t>(*buf_size_ptr);
+	  boSize[i] = buf_size;
+	  
+	  boHandles[i] = xclGetHostBO(devHdl,buf_addr,buf_size);
+	  bos[i] = xclMapBO(devHdl,boHandles[i],true);
+	  ffi_arg_values[i] = &bos[i];
+	  bo_list.emplace_back(i);
+	} else {
+	  ffi_arg_values[i] = &args_from_host[args[i].offset/4];
+	}
+      }
+    
+    }
+
+    // Original PS Kernel implementation
+    // Check for call signature of only 2 arguments and 2nd argument has name os ops
+    if((args.size()==2 && args[1].name.compare("ops")==0) || args.empty()) {
+      kernel_return = old_kernel(&args_from_host[1],&ops);
+    } else {
+      ffi_call(&cif,FFI_FN(kernel), &kernel_return, ffi_arg_values);
+    }
     args_from_host[1] = (uint32_t)kernel_return;
+
+    // Unmap Buffers
+    if((args.size()==2 && args[1].name.compare("ops")==0) || args.empty()) {
```

Comment:
It might be better if the args.empty() is before the other arguments.  It make the code easier to read.